WordCamp Central: Add global sponsors "block" to homepage

coreymckrill opened the pull request:
Swaps out the Latest Tweets column on the WordCamp Central [homepage](https://central.wordcamp.org) for a randomized list of three Gold-level sponsors. The seed for the randomize function changes every hour.

The code that outputs the sponsor list is structured similarly to WordCamp's Sponsor block, so that it would be easier to convert this into an actual block in the future.

Fixes #306

**Screenshot (Updated):**

![widget-logos](https://user-images.githubusercontent.com/916023/72482205-c4e06200-37b1-11ea-9ef1-e3a88652bd48.jpg)

~Right now we are lucky that all of the Gold-level sponsors have logos with similar aspect ratios. However, I think further design thinking is needed here about how best to accommodate one or more logos that aren't necessarily all in landscape orientation. @karmatosed~

~Also, does the heading make sense, given the link to the rest of the sponsors at the bottom? Should that link be there? @andreamiddleton~

**To test (Updated):**

* Load the WordCamp Central homepage in your sandbox.
* The widgets should already be there, I think, because I added them via my sandbox.
* Ensure that the third column below the hero contains Gold-level sponsors' logos, plus a link to the Global Sponsorship page.
* Try reloading the page a couple of times and make sure the order of the logos changes.

**Pre-deploy**

* [ ] Receive and confirm square logo assets from each of the Gold-level global sponsors. Update the widgets accordingly.

coreymckrill committed:
Multi Event Sponsors: Add proto-block for outputting a list of sponsors

WordCamp Central needs to show a list of sponsors on the homepage. This
was written as the beginnings of a block in case such a list becomes
useful elsewhere on the site at a future point.

coreymckrill committed:
gitignore: Add entry for uploads directory

coreymckrill committed:
Multi Event Sponsors: Ensure block initially retrieves all sponsor posts

It was defaulting to only 10 posts, so when randomized, sometimes the
final list was missing some posts.

coreymckrill committed:
WordCamp Central: Replace homepage tweets with sponsor logos

Rough draft, with a list of 3 logos at a time.

This hides the tweets column and related functionality, rather than
replacing it, in case the sponsors list changes or is removed in
future iterations of the program.

coreymckrill committed:
Multi Event Sponsors: phpcs cleanup on bootstrap

karmatosed commented:
Just one note, would it be possible to have 'needs design feedback' as a label if this just needs feedback? That's the flow we use on trac and other GitHub repos and might enable easier feedback.

Assuming this just needs design feedback, for me this works but as you note for now the logos fit. I don't think 'some of our sponsors' is the best copy but I will leave that to others to refine.

ryelle commented:
@karmatosed  Done— we now have [Needs Design Feedback](https://github.com/WordPress/wordcamp.org/labels/%5BStatus%5D%20Needs%20Design%20Feedback) & [Needs Design](https://github.com/WordPress/wordcamp.org/labels/%5BStatus%5D%20Needs%20Design) 👍

ryelle commented on public_html/wp-content/plugins/multi-event-sponsors/views/block-sponsor.php:
```diff
@@ -0,0 +1,26 @@
+/** @var array   $attributes */
+/** @var WP_Post $sponsor */
+
+setup_postdata( $sponsor ); // This is necessary for generating an excerpt from content if the excerpt field is empty.
```
Is this used in this block?

ryelle commented on public_html/wp-content/plugins/multi-event-sponsors/views/block-sponsor.php:
```diff
@@ -0,0 +1,26 @@
+namespace WordCamp\Multi_Event_Sponsors\Block;
+
+use WP_Post;
+use function WordCamp\Blocks\Components\{ render_featured_image, render_item_title, render_item_content, render_item_permalink };
```
Are all of these functions needed? I think you can just include `render_featured_image`.

coreymckrill replied:
Per [this comment](https://github.com/WordPress/wordcamp.org/issues/306#issuecomment-573198287), we're going to try using some other elements besides just the logo, but I'll be sure to remove the unused ones when the display is finalized.

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/js/central.js:
```diff
@@ -54,7 +54,7 @@ var WordCampCentral = ( function( $ ) {
 		try {
 			toggleNavigation();
-			populateLatestTweets();
+			//populateLatestTweets();
```
Why not remove the whole function?

coreymckrill replied:
91c2d44b

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/template-home.php:
```diff
@@ -105,7 +107,23 @@
+				'region_id'   => 'all',
+				'level_id'    => 3040794,
+				'image_align' => 'center',
+				'image_width' => 300,
```
This should probably be `600` so that it displays correctly on retina screens, the logos look fuzzy on my laptop. If you do, you'll want to add `height: auto` to the CSS for the `img` itself, otherwise it skews.

coreymckrill replied:
🤔 I assumed that the output of `render_featured_image` would include a `srcset` attribute, since it's using `get_the_post_thumbnail` under the hood. But you're right that it's not doing it in this case. Maybe my assumption is wrong? Looking into it.

ryelle replied:
yeah, that's interesting— i wonder if it's a theme-support thing, since this is based on 2010?

coreymckrill replied:
After some digging, I suspect it's because things go through multiple layers of `wp_kses_post`. Astonishingly, `srcset` and `sizes` are not considered valid attributes on the `img` tag with that set of allowed tags...

coreymckrill replied:
Looks related: https://core.trac.wordpress.org/ticket/29807

coreymckrill replied:
I guess for now we can just double the set width 😕

coreymckrill replied:
8f1e1c7e

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/template-home.php:
```diff
@@ -105,7 +107,23 @@
+			<a href="<?php echo esc_attr( get_permalink( get_page_by_path( 'global-community-sponsors' ) ) ); ?>" class="more">More sponsors &rarr;</a>
+		</div>  <!-- .wc-global-sponsors -->
+
+		<!--<div class="wc-tweets last">
```
Can we delete this code, rather than commenting it out?

coreymckrill replied:
91c2d44b

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/style.css:
```diff
@@ -1203,6 +1204,17 @@ li:after { content: "."; display: block; height: 0; clear: both; visibility: hid
+/* !Featured Sponsor */
+.wordcamp-mes-sponsors > li {
+	width: 100%;
+	/*min-height: 300px;
```
And here too— if we need to bring back the code, we can grab it from revision history.

coreymckrill replied:
Fair point, I can remove it all.

coreymckrill replied:
91c2d44b

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/style.css:
```diff
@@ -1203,6 +1204,17 @@ li:after { content: "."; display: block; height: 0; clear: both; visibility: hid
+	flex-direction: column;
+	justify-content: center;*/
+}
+.wordcamp-mes-sponsors .wordcamp-mes-sponsor__logo.align-center {
```
I think this is redundant, since the blocks CSS is output on all sites. I see this style already applied with `.wordcamp-image__featured-image-container.align-center`.

coreymckrill replied:
Good catch. be693910

ryelle reviewed (changes requested):
Good idea about making this the start of a block 👍 I have a few suggestions, nothing major.

I did notice that none of the logos I was looking at had alt text. I think that's something we can fix in the media library, so while these are being updated, we should make sure to do that (can probably just say the company name).

coreymckrill committed:
WordCamp Central: Remove unused latest tweets code

coreymckrill committed:
Remove redundant CSS rule

karmatosed commented:
@ryelle thanks 🙌

coreymckrill committed:
WordCamp Central: Use larger sponsor logos for retina support

The `srcset` and `sizes` attributes are getting stripped from the logo
image tags, probably because of this:
https://core.trac.wordpress.org/ticket/29807

So this is a workaround where we just use an image that's double the
size of the column it's going in.

andreamiddleton commented:
I'd like to change the title of the widget to "Gold Global Sponsors," please, and... is there a technical or design reason that it's preferable to show multiple sponsors in the widget, rather than a rotating display of one sponsor at a time? That's how we had been imagining it, but if there's a strong reason against doing it that way, I'm open to hearing about it! :)

ryelle reviewed (commented):
Code looks good now, though it sounds like the design/content might change, so I can give this another look over once it's ready to merge 👍

coreymckrill commented:
@andreamiddleton 

> is there a technical or design reason that it's preferable to show multiple sponsors in the widget, rather than a rotating display of one sponsor at a time? That's how we had been imagining it, but if there's a strong reason against doing it that way, I'm open to hearing about it! :)

There's some discussion about a rotating display ("slider") here: #306 

Basically, sliders/carousels are considered kind of an anti-pattern. That being said, if we're going to [try switching to just one logo at a time instead of a list of three](https://github.com/WordPress/wordcamp.org/issues/306#issuecomment-573198287) anyway, the slider functionality does have the added benefit of eventually displaying all of the gold sponsor logos on a single page load. If that's an important factor, we can explore adding the slider part. It shouldn't be hard to implement since we already have one further up on the page.

coreymckrill committed:
Change column heading to Gold Global Sponsors

coreymckrill commented:
Here's what a cycling logo would look like. Not technically "sliding" since the transition is a fade, but that's the only transition we can do while using the slideshow script that's already loaded on the page for the WordCamp slideshow at the top.

![logo-cycle](https://user-images.githubusercontent.com/916023/72377489-6c7f6680-36c4-11ea-8000-32e0cfc8b717.gif)

andreamiddleton commented:
That is what I asked for, and I see some problems with it, as you probably also do. :)

Is it possible to use specific files for this, rather than the logo that's a featured image on the global sponsor's listnig? We've had at least one global sponsors offer to create a special logo for this placement, with a concise call to action below the logo, and would like to offer that benefit to all Global Golds. 

Presumably if we're specifying a logo size, that will help resolve the "swimming in white space" challenge of fade/rotating logos idea?

coreymckrill commented:
A logo with a different aspect ratio (probably square) would definitely look better in that space.

Adding a separate input/file upload for this is not trivial, but doable. Is there anywhere else we would potentially use this alternate logo, or would it be only on the homepage? I worry that this would be adding a lot of complexity for something that was intended as just a small change.

Another alternative might be to put the Gold Sponsor logos in a row _beneath_ the three columns, that spans the entire content width. Then the landscape-oriented logos would have more space, could all be shown at once, and the height of the logo area would only need to accommodate one or two rows of logos.

coreymckrill commented:
Here's what it might look like with the logos beneath the columns:

![global-sponsors-footer](https://user-images.githubusercontent.com/916023/72467889-c64d6280-3790-11ea-8789-f43d4c5ff764.jpg)

iandunn commented:
Adding another uploader for this seems like overkill, personally. 

@andreamiddleton, what are you concerned about with the proposed approach?

andreamiddleton commented:
Two concerns are motivating me here:

The general vision had been for one sponsor to be displayed at a time, even if it's just a short amount of time, which is preferable to the sponsoring companies of course. :) 

And then, in discussion of this benefit with one global sponsor, they asked whether they could add a 3-5 word tagline/call to action under their logo for this particular case, which the global sponsorship stewards (me, Courtney, and Kevin) approved. I don't know whether they were hoping to use that logo treatment on all WordCamp sites -- I suspect not -- so we'd need a unique logo for this particular widget. 

I did not convey this information very clearly, or possibly at all, and for that I'm really sorry. :(

iandunn commented:
🤔 

What if we output a hidden `div` with each sponsor's existing image/text when PHP renders the page, and then use JS to toggle through them every `n` seconds?

That way only 1 would be visible at a time, but we wouldn't have to deal w/ the slideshow script. IIRC jQuery is already on the page, and could maybe provide a transition, even it's it's just a quick `hide()` and then `show()`.

( we can't randomize in PHP b/c of static page caching, not sure if that's been mentioned anywhere explicitly yet )

andreamiddleton commented:
That could work, I suppose, though I suspect that a sponsor would prefer to have the tagline/call to action in a font that works best with their logo, so it could get complicated to serve up that text for them. I guess I didn't realize that it would be tricky to pull the images from this widget from the media library rather than the global sponsor listing? Or maybe it wouldn't have been tricky if I had mentioned the idea beforehand. 😒

iandunn commented:
>  I guess I didn't realize that it would be tricky to pull the images from this widget from the media library rather than the global sponsor listing? 

Er, yeah, that's a great clarification. I think we maybe got off track thinking that this needed to be 100% automated. 

We could just upload those images manually, and manually create an HTML Widget (or an Image widget in a custom sidebar, or whatever). It's only 4 companies, and they only update once a year.

Then, we can use the JS bit to show/hide them?

andreamiddleton commented:
> It's only 4 companies, and they only update once a year.

Yes. 😄 It's ok if we have to do a little manual labor to set the right logos, here I think. The pool of images will naturally be small, and change very infrequently.

iandunn commented:
Might even be able to use this to play around w/ FSE if the Gutenberg plugin supports it.

coreymckrill committed:
Add widget area to homepage

coreymckrill commented:
Here's what it would look like with the column as a widget area, and each sponsor as an image widget, with a (hopefully square) logo image and a caption underneath as the short call to action.

![widget-logos](https://user-images.githubusercontent.com/916023/72482205-c4e06200-37b1-11ea-9ef1-e3a88652bd48.jpg)

iandunn commented:
👍 

> caption underneath as the short call to action

It sounds like [they wanted the blurb to be a part of the image](https://github.com/WordPress/wordcamp.org/pull/314#issuecomment-574881599), so that they can use specific fonts, etc.

That shouldn't change anything AFAICT, just FYI.

I guess the caption actually is nice, though, b/c it gives people the option of putting it in the image, or in text.

coreymckrill committed:
Add classes to widget area list container

coreymckrill committed:
Update description of sponsors widget area

coreymckrill committed:
Randomize the order of sponsor widget area widgets

coreymckrill committed:
Turn sponsor widge area into a slideshow

coreymckrill committed:
Tweak sponsor widget area styles

coreymckrill committed:
Remove unused Multi Event Sponsor block code

coreymckrill committed:
Revert remaining changes in Multi Event Sponsors

coreymckrill commented:
Marking this as "On Hold" until we receive all the correct logo assets. In the mean time, it can still be reviewed.

ryelle commented on public_html/wp-content/themes/wordcamp-central-2012/style.css:
```diff
@@ -1100,6 +1100,7 @@ li:after { content: "."; display: block; height: 0; clear: both; visibility: hid
 }
 #wc-content-blocks .wc-news h3 { background-position: 0 -71px }
+#wc-content-blocks .wc-global-sponsors h3 { background-position: 0 -147px; }
 #wc-content-blocks .wc-tweets h3 {
```
There are still a few styles related to twitter (`wc-tweet`) in here, those could also be removed

coreymckrill replied:
Good catch. Fixed with 09cade7f

ryelle commented:
Looks good on desktop, but smaller screens are a little janky.

When it switches to a single column (660px), the images overlap the footer:

<img width="455" alt="Screen Shot 2020-01-17 at 9 36 05 AM" src="https://user-images.githubusercontent.com/541093/72621243-e1de7780-390e-11ea-9b25-140f0bfe7abe.png">

And when the columns are less than 300px (screen width about 960px), the logos get cut off on the right:

<img width="196" alt="Screen Shot 2020-01-17 at 9 47 04 AM" src="https://user-images.githubusercontent.com/541093/72621363-1e11d800-390f-11ea-857f-5989eec2d830.png">

coreymckrill committed:
Remove extraneous tweet-related styles

coreymckrill committed:
Fix responsive layout issues with wiget

* Override the hard-coded inline width on the image widget caption so
  that the logos don't get cut off when the column width is less than
  300px
* Set a hard-coded height on the column so that the images don't overlap
  the footer on small screens. This is necessary because the cycle
  plugin relies on absolute positioning, and a simple clearfix won't
  do the trick.

coreymckrill commented:
Ok, the small screen issues should be fixed with 15e5700a

coreymckrill committed:
Don't break sponsor taglines mid-word
